core/src/main/java/io/opentelemetry/android/export/BufferedDelegatingExporter.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.export
+
+import io.opentelemetry.sdk.common.CompletableResultCode
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * An in-memory buffer delegating signal exporter that buffers signal in memory until a delegate is set.
+ * Once a delegate is set, the buffered signals are exported to the delegate.
+ *
+ * The buffer size is set to 5,000 by default. If the buffer is full, the exporter will drop new signals.
```

Comment:
It was arbitrarily picked. It was chosen to be high enough to handle most logging needs while not taking up too much space. It can absolutely be changed

Reply:
Seems like a reasonable default to me. 🍰

Reply:
I don't have a clear reference of how much memory each signal can take, mostly because it also depends on its attributes/general content, though iirc a simple proto-serialized span size in bytes can get to around 700 (I might be very wrong though because last time I checked was quite long ago), let's say it's 500 bytes per span for the sake of the example, then we might be talking about ~2.5 mb for 5000 of those, plus other ~2.5 mb for the logs, we might be talking about ~5 mb total in the worst case, though again it depends on the size of each signal and also on how accurately I remember these rough proto sizes and how they translate into runtime memory, so we should take this with a grain of salt. 

However, before getting into the weeds of how much memory is available for apps/OS version/hardware/etc, I'd like to take a step back and try to figure out for how long do we think the agent would need to use this mechanism of storing signals in memory? If we base it on the ANR issue that's the reason for these changes, it means that it would roughly take as long as it takes `android.os.storage.StorageManager.getAllocatableBytes` to execute, which I haven't measured but I'm guessing it should take less than a second, and, based on that assumption, I'm finding it difficult to see how 10000 signals can be created within that timeframe. I tend to err on the side of using as few resources as possible, so I'd prefer to choose a much lower number if possible. From a couple of local examples I've run (not real-world examples though) I think I've gotten at most ~20 signals created at app launch, considering it's not a real-world use case I'm conscious that the number might be higher, but 5000 seems a bit excessive. I might be missing something though, so I was wondering if you could share a bit more info on the kind of use-cases that we should look out for where this number might be needed?

Reply:
Your analysis matches our own. Our goal was to pick a value high enough that nobody is likely to hit it but low enough to have little impact on memory usage.

We've seen that the binder call can take up to several seconds to complete on a low hardware phone that is currently busy with background services so I think it's fair to say that 5s is around a common worst case (something like P99).

We don't mind setting it lower but the risk of dropping events without any warning seems potentially very very bad. It's worth noting that we do not allocate more space than needed (or rather the array list doesn't). It doubles in size when more space is needed so you'd be logging more than 2500 spans to get the full 5000 capacity. And if you're logging that much you would likely trade 2.5MB (per exporter) of memory over dropping them I'd assume.

That being said, this value could definitely be lowered if you feel like it's absurdly high but I wanted to outline our thinking here. Let us know how you want to proceed.

Reply:
Fair points. I wasn't aware that this call could take up to 5 seconds to execute in some cases, and I was mostly curious to know about some use cases that would require that amount of signals within that timeframe. However, as you mentioned, the number is high enough for it to be difficult to get hit, and also, the memory isn't allocated upfront anyway, so it should be fine to leave it as is. Thanks for the explanation.

Reply:
Let's also not forget that we can change this default in a future PR....there's nothing holding us to this number. 🙃